dev-docs/RFCs/v8.0/carto-v9-rfc.md

```diff
@@ -0,0 +1,222 @@
+- No easy way for the developer to use a custom layer for rendering.
+- No types to enforce correct options.
+
+We are working right now in v9, so I think it's a great time to address those changes. Especially because v9 will expose types at the package roots.
+
+## Background
+
+The current implementation of `CartoLayer` is as a `CompositeLayer` which encapsulates all the data loading via the CARTO API, and depending on the particular configuration of props renders the data using one of the following layers:
+
+- `CartoTileLayer`
+- `H3TileLayer`
+- `MVTLayer`
+- `RasterTileLayer`
+- `QuadbinTileLayer`
+
+The standard pattern in deck.gl is for a `data` prop of `Layer` to accept the actual data to be visualized, or a Promise that resolves to the data. The `MVTLayer` additionally [accepts TileJSON](https://deck.gl/docs/api-reference/geo-layers/mvt-layer#data) as input, as do all the CARTO Layers.
+
+## API background
+
+Without getting into the details too much, using the current `CartoLayer` with the CARTO works in the following way:
+
+1. User configures `CartoLayer` with a set of props
+2. `CartoLayer` uses these props to construct a URL for a Map Instantiation request
+3. `CartoLayer` performs a Map Instantiation fetch request and obtains URLs for data endpoints, in different formats: TileJSON, GeoJSON, JSON etc
+4. `CartoLayer` is [hardcoded](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/layers/carto-layer.ts#L263) to only use TileJSON, so the TileJSON URL is fetched
+5. `CartoLayer` [decides](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/api/layer-map.ts#L203-L223) which Layer to use for rendering and passes the TileJSON payload to that layer in the `data` prop
+
+## Proposal
+
+At a high level we want to achieve an API which makes it clearer to the user what is going on, and allows for better extensibility. We also want to separate the options used for loading the data from the props which style the layer.
+
+To do so, the interaction with the CARTO API will be done via a set of typed helper classes which aid the user in obtaining the TileJSON needed for a particular layer.
+
+```javascript
+import CartoQuerySource, CartoQuerySourceOptions, CartoTileLayer from '@deck.gl/carto';
+
+// Global options
+setDefaultCredentials({accessToken: 'XXXX'});
+
+// All properties in one place
+const options: CartoQuerySourceOptions = {
+  connectionName: 'bigquery',
+  sqlQuery: 'select * from carto.tables.example'
+}
+
+// Returns a Promise which resolves to TileJSON payload. Can optionally await if needed
+const data = CartoQuerySource(options);
+
+const layer = new CartoTileLayer({
+  data,
+
+  // ...style props
+});
+```
+
+## Class names
+
+The goal is to have descriptive names for classes and options, even if this means they are more verbose.
+
+| Source class                | Layer                   |
```

Comment:
What about removing the `tile` suffix? Everything is a tile layer and the idea is to continue like that

Reply:
Yes, it's a pain having it everywhere. I've removed it, but I'm wary of keeping the generic `CartoLayer` name as people will just expect it to continue to work as before. `CartoVectorLayer`?